src/main/java/com/amazon/opendistroforelasticsearch/knn/plugin/script/KNNScoringUtil.java

```diff
@@ -0,0 +1,79 @@
+package com.amazon.opendistroforelasticsearch.knn.plugin.script;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+
+public class KNNScoringUtil {
+    private static Logger logger = LogManager.getLogger(KNNScoringUtil.class);
+
+    public static float l2Squared(float[] queryVector, float[] inputVector) {
+        float squaredDistance = 0;
+        for (int i = 0; i < inputVector.length; i++) {
+            float diff = queryVector[i]-inputVector[i];
+            squaredDistance += diff * diff;
+        }
+        return squaredDistance;
+    }
+
+    public static float cosinesimilOptimized(float[] queryVector, float[] inputVector, double normQueryVector) {
+        double dotProduct = 0.0f;
+        double normInputVector = 0.0f;
+        if (normQueryVector == -1) {
+            throw new IllegalStateException("Normalized query vector cannot be negative");
+        }
+        for (int i = 0; i < queryVector.length; i++) {
+            dotProduct += queryVector[i] * inputVector[i];
+            normInputVector += inputVector[i] * inputVector[i];
+        }
+        double normalizedProduct = normQueryVector * normInputVector;
+        try {
+            return (float) (dotProduct / (Math.sqrt(normalizedProduct)));
+        } catch(ArithmeticException ex) {
+            logger.debug("Possibly Division by Zero Exception. Returning min score to put this result to end. " +
+                    "Current normalized product " + normalizedProduct);
+            return Float.MIN_VALUE;
+        }
+    }
+
+    public static float cosinesimil(float[] queryVector, float[] inputVector) {
+        double dotProduct = 0.0f;
+        double normQueryVector = 0.0f;
+        double normInputVector = 0.0f;
+        for (int i = 0; i < queryVector.length; i++) {
+            dotProduct += queryVector[i] * inputVector[i];
+            normQueryVector += queryVector[i] * queryVector[i];
+            normInputVector += inputVector[i] * inputVector[i];
```

Comment:
Recomputing the query vector of every invocation will add unnecessary computation, as it can be computed once and passed it, like in the function above

Reply:
I added two versions of calculating cosine similarity score. This is unoptimized version of cosine similarity function which would be exposed to customer to call the function directly.